ruby/tests/service_test.rb

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/ruby
+
+require 'google/protobuf'
+require 'service_test_pb'
+require 'test/unit'
+require 'json'
+
+class ServiceTest < Test::Unit::TestCase
+  def setup
+    @test_service = Google::Protobuf::DescriptorPool.generated_pool.lookup('service_test_protos.TestService')
+    @deprecated_service = Google::Protobuf::DescriptorPool.generated_pool.lookup('service_test_protos.DeprecatedService')
+    @test_service_methods = @test_service.methods.to_h { |method| [method.name, method] }
+  end
+
+  def test_lookup_service_descriptor
+    assert_kind_of Google::Protobuf::ServiceDescriptor, @test_service
+    assert_equal 'service_test_protos.TestService', @test_service.name
+  end
+
+  def test_file_descriptor
+    assert_kind_of Google::Protobuf::FileDescriptor, @test_service.file_descriptor
+    assert_equal 'service_test.proto', @test_service.file_descriptor.name
+  end
+
+  def test_method_iteration
+    @test_service.methods.each { |method| assert_kind_of Google::Protobuf::MethodDescriptor, method }
+    assert_equal %w(UnaryOne UnaryTwo), @test_service.methods.map { |method| method.name }.first(2)
+  end
+
+  def test_service_options
+    assert @deprecated_service.options.deprecated
+    refute @test_service.options.deprecated
+  end
+
+  def test_service_options_extensions
+    # The ruby API doesn't actually support extensions it seems yet, but we can bodge it for this
+    # test by abusing to_json (but only for the UPB implementation of to_json it seems - the java
+    # one doesn't do this).
+    omit('JRuby implementation cannot display option extensions') if defined?(::JRUBY_VERSION)
+
+    options_hash = JSON.parse(@test_service.options.to_json)
```

Comment:
So the actual utility of this for the original use-case in https://github.com/protocolbuffers/protobuf/issues/14891 (getting RPC options) might be somewhat dubious without support for being able to get message extensions (without resorting to `#to_json`). I guess that can be tackled in a subsequent PR